src/core/reference/include/openvino/reference/reduce_l1.hpp

```diff
@@ -7,33 +7,55 @@
 #include <cmath>
 #include <numeric>
 
-#include "ngraph/shape_util.hpp"
-#include "openvino/reference/utils/coordinate_transform.hpp"
+#include "openvino/reference/sum.hpp"
+#include "openvino/reference/utils/type_util.hpp"
+#include "shape_util.hpp"
 
 namespace ov {
 namespace reference {
-template <typename T>
-void reduce_l1(const T* arg, T* out, const Shape& in_shape, const AxisSet& reduction_axes) {
-    constexpr bool dont_keep_dims_in_output = false;
-    OPENVINO_SUPPRESS_DEPRECATED_START
-    const auto out_shape = ngraph::reduce(in_shape, reduction_axes, dont_keep_dims_in_output);
-    std::fill(out, out + shape_size(out_shape), T(0));
+
+namespace func {
+template <class T, typename std::enable_if<std::is_unsigned<T>::value>::type* = nullptr>
+constexpr T abs(const T num) {
+    return num;
+}
+
+template <class T, typename std::enable_if<std::is_signed<T>::value || ov::is_floating_point<T>()>::type* = nullptr>
+T abs(const T num) {
+    return std::abs(num);
+}
+}  // namespace func
+
+/**
+ * @brief Reference implementation of ReduceL1 operator.
+ *
+ * @param in             Input iterator to data.
+ * @param out            Output iterator to results.
+ * @param in_shape       Input shape.
+ * @param reduction_axes Axes on which reduction is applied.
+ */
+template <class InputIt, class OutputIt>
+void reduce_l1(InputIt in, OutputIt out, const Shape& in_shape, const AxisSet& reduction_axes) {
+    using T = typename std::iterator_traits<OutputIt>::value_type;
+    static_assert(std::is_same<typename std::iterator_traits<InputIt>::value_type, T>::value,
+                  "Assume in and out same type.");
```

Comment:
The goal is to migrate to new API, the refactor is optional on this stage unless is relative easy.

In cpp the recommended way is use iterators over raw pointers when same iteration over data are done.
The iterators give more flexibility, as iterator can be e.g. using iterator adaptor. 
In reduce there could be some `stride iterator` adaptor which  could work similar as coordinate iterator but instead generate coordinate and calculate product just shift iterator to next element. Then instead loop some stl algo can be used which can impact on reducing code size and some performance as less multiplication will be used.
Using iterator also avoid do pointer arithmetic in explicit way, which is also not recommended to do.

I've check impact on binary size and here this change not affect it (but I've seen on some other implementations there is some change which increase size, then will leave pointer until root cause)

Regarding const T* can be achieved by providing const iterator or iterator from const container, then if function want change input data there will be an error.

I'm not sure if this approach can be applied to all reference but if there is no size penalty it gives more flexibility as template implementation.

Reply:
> I'm not sure if this approach can be applied to all reference but if there is no size penalty it gives more flexibility as template implementation.

There shouldn't be any size impact until new usages with other in/out template types.
Of course, I agree that usage of the iterators inside the reference implementation is good, but in fact the raw pointers are still passed to the function, because this is what `tensor.data()` returns.